### PR TITLE
fix(msa): ensure schema ids are valid

### LINKED ISSUE #506: Add validation schema grant validation
## Description
Ensure that schema grants are validated when added.

## Acceptance Criteria
- [ ] any schema added should be less or equal to the latest [schema id](https://github.com/LibertyDSNP/frequency/blob/main/pallets/schemas/src/lib.rs#L158).

### PR DESCRIPTION
**Pull request by enddynayn**
# Goal
Ensure that all schema ids granted to
providers are valid before storing.

Remove duplicate validation check
for exceeding max allowed schema grants.

Closes #506

### CODE REVIEW

File: pallets/msa/src/lib.rs [hunk 1]
```diff
@@ -610,6 +609,21 @@ impl<T: Config> Pallet<T> {
 		})
 	}
 
+	/// Check that schema ids are all valid
+	pub fn ensure_schema_ids_are_valid(
+		schema_ids: BoundedVec<SchemaId, T::MaxSchemaGrants>,
+	) -> DispatchResult {
+		let schema_count = T::SchemaProvider::get_schema_count();
+
+		if_std! {
```
Comment (enddynayn) [inline pallets/msa/src/lib.rs:618]: remove

File: pallets/msa/src/tests.rs [hunk 2]
```diff
@@ -18,9 +18,12 @@ use frame_support::{
 	assert_err, assert_noop, assert_ok,
 	weights::{DispatchInfo, GetDispatchInfo, Pays, Weight},
 };
+// use pallet_schemas::SchemaCount;
 use sp_core::{crypto::AccountId32, sr25519, Encode, Pair};
 use sp_runtime::{traits::SignedExtension, MultiSignature};
 
+// fn set_schema_count(number: u16) {}
```
Comment (enddynayn) [inline pallets/msa/src/tests.rs:25]: remove

File: pallets/msa/src/tests.rs [hunk 3]
```diff
@@ -18,9 +18,12 @@ use frame_support::{
 	assert_err, assert_noop, assert_ok,
 	weights::{DispatchInfo, GetDispatchInfo, Pays, Weight},
 };
+// use pallet_schemas::SchemaCount;
```
Comment (enddynayn) [inline pallets/msa/src/tests.rs:21]: remove

File: pallets/schemas/src/lib.rs [hunk 4]
```diff
@@ -310,6 +311,15 @@ impl<T: Config> SchemaProvider<SchemaId> for Pallet<T> {
 	fn get_schema_by_id(schema_id: SchemaId) -> Option<SchemaResponse> {
 		Self::get_schema_by_id(schema_id)
 	}
+
+	fn get_schema_count() -> SchemaId {
+		Self::get_schema_count()
+	}
+
+	fn is_valid_schema_id(schema_id: SchemaId) -> bool {
```
Comment (enddynayn) [inline pallets/schemas/src/lib.rs:319]: remove

File: pallets/messages/src/mock.rs [hunk 5]
```diff
@@ -193,6 +193,14 @@ impl SchemaProvider<u16> for SchemaHandler {
 			payload_location: PayloadLocation::default(),
 		})
 	}
+
+	fn is_valid_schema_id(_schema_id: SchemaId) -> bool {
```
Comment (enddynayn) [inline pallets/messages/src/mock.rs:197]: remove

File: pallets/msa/src/lib.rs [hunk 6]
```diff
@@ -610,6 +610,21 @@ impl<T: Config> Pallet<T> {
...
```
Comment (enddynayn) [inline pallets/msa/src/lib.rs:619]: remove

File: pallets/msa/src/lib.rs [hunk 7]
```diff
@@ -671,6 +669,17 @@ impl<T: Config> Pallet<T> {
 		})
 	}
 
+	/// Check that schema ids are all valid
+	pub fn ensure_schema_ids_are_valid(
```
Comment (enddynayn) [inline pallets/msa/src/lib.rs:673]: SchemaValidator interface

```
are_schema_ids_valid(Iterator | Vec) -> bool
```

File: pallets/schemas/src/lib.rs [hunk 8]
```diff
@@ -266,11 +272,6 @@ pub mod pallet {
 			Ok(schema_id)
 		}
 
-		/// Retrieve latest schema id via total count of schemas on chain
-		pub fn get_latest_schema_id() -> Option<SchemaId> {
```
Comment (enddynayn) [inline pallets/schemas/src/lib.rs:270]: I could not find where this is being used so removed it.

**Review by wilwade — APPROVED** [review]
👍 If you have to rebase again, be nice to remove the `.git` from the lock file that something keeps trying to add in, but not a big issue as there is a real change to the lock file.

**Pull request merged**